File: library/src/main/java/no/wtw/android/restserviceutils/RestServiceAPI.java
package no.wtw.android.restserviceutils;

import android.content.Context;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.util.Log;

import org.springframework.http.HttpAuthentication;
import org.springframework.http.HttpBasicAuthentication;
import org.springframework.http.HttpStatus;
import org.springframework.http.client.ClientHttpRequestInterceptor;
import org.springframework.web.client.RestTemplate;

import java.util.Arrays;

import no.wtw.android.restserviceutils.exceptions.RestServiceException;

public abstract class RestServiceAPI<S> {

    private static final String TAG = RestServiceAPI.class.getSimpleName();

    public void setDefaultInterceptor() {
        ClientHttpRequestInterceptor i = new RestServiceRequestInterceptor(this);
        getRestTemplate().setInterceptors(Arrays.asList(i));
    }

    public void setDefaultErrorHandler() {
        getRestTemplate().setErrorHandler(new RestServiceErrorHandler());
    }

    public boolean isOnline() {
        ConnectivityManager cm = (ConnectivityManager) getContext().getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo activeNetworkInfo = cm != null ? cm.getActiveNetworkInfo() : null;
        return activeNetworkInfo != null && activeNetworkInfo.isAvailable() && activeNetworkInfo.isConnected();
    }

    /**
     * Convenience method that throws an exception if network connection is not available.
     * May be overridden by sub class to provide better information on what to do in offline cases.
     *
     * @throws RestServiceException
     */
    public void checkNetwork() throws RestServiceException {
        if (!isOnline())
            throw new RestServiceException(HttpStatus.SERVICE_UNAVAILABLE, "Network unavailable");
    }

    public abstract RestTemplate getRestTemplate();

    public abstract Context getContext();

    public HttpAuthentication getAuthentication() {
        return new HttpBasicAuthentication("", "");
    }

    public <T, C extends Call<S, T>> T call(C call) throws RestServiceException {
        try {
            checkNetwork();
            return call.execute(getService());
        } catch (Exception e) {
            if (e.getMessage() != null)
                Log.e(TAG, e.getMessage());
            throw RestServiceException.getInstance(e);
        }
    }

    protected abstract S getService();

    public interface Call<S, T> {
        T execute(S service);
    }

}

File: library/src/main/java/no/wtw/android/restserviceutils/RestServiceAPI.kt
package no.wtw.android.restserviceutils

import android.content.Context
import android.net.ConnectivityManager
import com.google.gson.Gson
import no.wtw.android.restserviceutils.exceptions.RestServiceException
import no.wtw.android.restserviceutils.resource.JsonEncodedQuery
import no.wtw.android.restserviceutils.resource.Link
import okhttp3.OkHttpClient
import java.util.*

abstract class RestServiceAPI {

    /**
     * Convenience method that throws an exception if network connection is not available.
     * May be overridden by sub class to provide better information on what to do in offline cases.
     *
     * @throws RestServiceException
     */
    @Throws(RestServiceException::class)
    fun assertIsOnline() {
        if (!isOnline)
            throw RestServiceException(HttpStatusCode.SERVICE_UNAVAILABLE, "Network unavailable")
    }

    val isOnline: Boolean
        get() {
            val cm = getContext().getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
            val activeNetworkInfo = cm.activeNetworkInfo
            return activeNetworkInfo != null && activeNetworkInfo.isAvailable && activeNetworkInfo.isConnected
        }

    abstract fun getContext(): Context
    abstract fun getClient(): OkHttpClient
    abstract fun getGson(): Gson
    open fun getConnectTimeout() = 60000
    open fun getReadTimeout() = 60000

}

fun <T> Link<T>.httpGet(api: RestServiceAPI): T = this.httpGet(api.getClient(), api.getGson())
fun <T> Link<T>.httpGet(api: RestServiceAPI, queryParams: Map<String, String>): T = this.httpGet(api.getClient(), api.getGson(), queryParams)
fun <T> Link<T>.httpGet(api: RestServiceAPI, query: JsonEncodedQuery): T = this.httpGet(api.getClient(), api.getGson(), query)
fun <T> Link<T>.httpDelete(api: RestServiceAPI): Unit = this.httpDelete(api.getClient(), api.getGson())
fun <T> Link<T>.httpPut(api: RestServiceAPI, body: T): T = this.httpPut(api.getClient(), api.getGson(), body)
fun <T> Link<T>.httpPost(api: RestServiceAPI, body: Any?): T = this.httpPost(api.getClient(), api.getGson(), body)
